llvm/include/llvm/Analysis/AliasSetTracker.h

```diff
@@ -307,7 +173,7 @@ class AliasSetTracker {
   BatchAAResults &AA;
   ilist<AliasSet> AliasSets;
 
-  using PointerMapType = DenseMap<AssertingVH<Value>, AliasSet::PointerRec *>;
+  using PointerMapType = DenseMap<MemoryLocation, AliasSet *>;
```

Comment:
I don't really get why this needs to be indexed by MemoryLocation. I'd still expect all MemoryLocations for the same Value to be in the same AliasSet, so it seems like the existing indexing should work fine...

Reply:
> I'd still expect all MemoryLocations for the same Value to be in the same AliasSet

I was unsure if this always holds. In case of full restrict support, I think the same pointer value is used for the two stores in the following snippet:

```
int * restrict r = p;
*p = 123;
*r = 456;
```

But they have different provenance and metadata, and ScopedNoAliasAA will report that they do not alias. But this snippet has undefined behavior (because of course, the accesses do alias).

I guess that BasicAA will quickly return MustAlias for a query with the same pointer Value regardless of LocationSize, metadata, etc.? Then indeed another AA can only say NoAlias in an illegal situation.

Reply:
> I don't really get why this needs to be indexed by MemoryLocation [...] it seems like the existing indexing should work fine...

I still tried this, see [this branch](https://github.com/brunodf-snps/llvm-project/tree/ast-pointer-map). The pointer map can be smaller, but this is traded for extra work in `AliasSetTracker::getAliasSetFor`, see commit 23443f3. It turns out that two memory locations for the same `UndefValue` **do** end up in different alias sets (see [this BasicAA logic](https://github.com/llvm/llvm-project/blob/92fc4b482f58cb243d982d572adb7a80ceab6945/llvm/lib/Analysis/BasicAliasAnalysis.cpp#L1472-L1475)) so I have to exclude them from the pointer map, which complicate the logic. On the plus side, a lots of tests can be reverted, see commit 92f01a2, since we can restore the "pointer values" output in the alias set dump, but I don't know if that is much of an argument.

Let me know if you consider the version on the branch an improvement, then I will include it in this pull request.

Reply:
I still realized there was a problem with commit 23443f3. When a memory location with an undef pointer value is added to an alias set, that alias set is not referenced from the pointer map entry of that pointer value (because the pointer value may appear in _multiple_ alias sets). But without the reference to the alias set from some access structure, the reference count of the set may reach zero, and the alias set may be dropped. (For similar reason, the reference count of an alias set is increased by 1 when at least one unknown instruction is registered in it.)

I solved this problem (on [the aforementioned branch](https://github.com/brunodf-snps/llvm-project/commits/ast-pointer-map/)) by keeping a collection with alias sets that contain undef pointer values. This keeps those alias sets referenced, so they are not dropped. It also allows to scan only those alias sets, when we are trying to find if a memory location with an undef pointer value is already registered or not.

Reply:
I'm confused. Yes, AA treats undef as noalias, but as far as I know AST specifically wants the undefs to be part of the same set, or at least that is how I interpret the comment at https://github.com/llvm/llvm-project/blob/5f254eb05566f5b400a212ae77117a9efd9019a1/llvm/lib/Analysis/AliasSetTracker.cpp#L346-L350. It seems like you are going out of your way to have undefs in different sets -- why?

More generally I would expect that all MemLocs that have the same pointer will end up in the same alias set, even if AA reports them as NoAlias (as implied by metadata or undef). After all they are *also* MustAlias, and treating them as such makes things a lot simpler.

Reply:
I agree that when using a map from pointer values, it is more logical to simply place MemLocs in an existing set of the pointer value, even if AA reports NoAlias for this set. I've updated the ast-pointer-map branch to this effect, in particular, see:
https://github.com/brunodf-snps/llvm-project/blob/9e9ebcd53aff011cff2b4d7ebdf439855806894b/llvm/lib/Analysis/AliasSetTracker.cpp#L312-L319

But be aware that in a program where AA indicates that two instruction using an undef pointer are not aliasing, _putting them in the same alias set is information loss_ (although it may not be relevant):

```
%val = load i32, ptr undef
...
store ptr null, ptr undef
``` 

So I would disagree with the cited code comment, or at least find it misleading:

>> Since alias(undef, undef) is NoAlias, mergeAliasSetsForPointer(undef, ...) will not find the the right set for undef, even if it exists. 

`mergeAliasSetsForPointer` is correctly indicating (with a `nullptr` result) that there is no set that the second memory location aliases with, according to AA. The existing set for the first memory location with an undef pointer is only a "right" set insofar that the implementation with a pointer map cannot accurately represent the situation. (But it may be a fine tradeoff to be inaccurate here.)

> It seems like you are going out of your way to have undefs in different sets -- why?

Well, I have roughly 2 objectives for AST with this patch: (1) avoid information loss and (2) minimal knowledge about AA or MemoryLocation internals.  So I am reluctant to add information loss for an implementation organization (the pointer map). But you seem to favor the pointer map for conceptual reasons too, so then I would propose what is on my ast-pointer-map branch.

Reply:
It is true that both NoAlias and MustAlias are correct results here and you can choose which one to pick, but my reasoning was that AA already picks NoAlias and so I attempted to represent the same thing in AST (when I say I strive for no information loss, I essentially mean that AST should construct alias sets consistent with AA results -- the motivating example is a case where AA is indicating NoAlias, and current AST is ending up with MayAlias due to implementation organization).

But I am fine with an implementation where all MemoryLocations for the same pointer Value are by construction in the same AliasSet, i.e. where the PointerMap stays. The discussion has shown that the impact is only for undef pointer values, and I expect the difference is of very little practical relevance (as you basically write, creating a new set for every undef pointer does not seem to have added value). So I have merged my ast-pointer-map branch into this PR now, so the combined set of changes that I propose is visible here, and we can move forward with that.